lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,683 @@
+
+    // Go through the port annotations of the tap module and generate a
+    // hierarchical path for each.
+    SmallVector<PortWiring, 8> portWiring;
+    SmallDenseMap<Attribute, unsigned, 2> memPortIdx;
+    portWiring.reserve(blackBox.portAnnos.size());
+    for (auto portAnno : blackBox.portAnnos) {
+      LLVM_DEBUG(llvm::dbgs() << "- Processing port " << portAnno.portNum
+                              << " anno " << portAnno.anno << "\n");
+      auto portName = getModulePortName(blackBox.extModule, portAnno.portNum);
+      auto cls = portAnno.anno.getAs<StringAttr>("class");
+      PortWiring wiring = {portAnno.portNum, {}, {}};
+
+      // Handle data taps on signals and ports.
+      if (cls == strings.referenceKeyClass) {
+        // Handle block arguments.
+        if (auto blockArg = tappedArgs.lookup(portAnno.anno)) {
+          auto parentModule = blockArg.getOwner()->getParentOp();
+          wiring.prefices = instanceGraph.getAbsolutePaths(parentModule);
+          wiring.suffix =
+              getModulePortName(parentModule, blockArg.getArgNumber())
+                  .getValue();
+          portWiring.push_back(std::move(wiring));
+          continue;
+        }
+
+        // Handle operations.
+        if (auto op = tappedOps.lookup(portAnno.anno)) {
+          // We currently require the target to be a wire.
+          // TODO: This should probably also allow wires?
+          auto wire = dyn_cast<WireOp>(op);
+          if (!wire) {
+            auto diag = blackBox.extModule.emitError(llvm::formatv(
+                "ReferenceDataTapKey on port {0} must be a wire", portName));
+            diag.attachNote(op->getLoc()) << "referenced operation is here:";
+            signalPassFailure();
+            continue;
+          }
+
+          // We currently require the target to be named.
+          // TODO: If we were to use proper cross-module reference ops in the IR
+          // then this could be anonymous, with ExportVerilog resolving the name
+          // at the last moment.
+          auto name = wire.nameAttr();
+          if (!name) {
+            auto diag =
+                wire.emitError("wire targeted by data tap must have a name");
+            diag.attachNote(blackBox.extModule->getLoc()) << llvm::formatv(
+                "used by ReferenceDataTapKey on port {0} here:", portName);
+            signalPassFailure();
+            continue;
+          }
+
+          wiring.prefices = instanceGraph.getAbsolutePaths(
+              wire->getParentOfType<FModuleOp>());
+          wiring.suffix = name.getValue();
+          portWiring.push_back(std::move(wiring));
+          continue;
+        }
+
+        // The annotation scattering must have placed this annotation on some
+        // target operation or block argument, which we should have picked up in
+        // the tapped args or ops maps.
+        signalPassFailure();
+        llvm::report_fatal_error(
```

Comment:
Can this be an emitOpError?  That will carry location information.

Reply:
Done!